File: davidson.py
import numpy as np
import argparse

def gramschmidt(X):

    d, n = X.shape
    m = min(d,n)
    R = np.zeros((m,n))
    Q = np.zeros((d,m))
    Q2 = np.zeros((d,n))

    for i in range(m):
        v = X[:,i]
        for j in range(i):
            R[j,i] = np.dot(Q[:,j].T,v)
            v -= R[j,i]*Q[:,j]
        R[i,i] = np.linalg.norm(v)
        Q[:,i] = v/R[i,i]
    R[:,m+1:] = np.dot(Q.T,X[:,m+1:])

    return Q,R

def diagonal_guess(nroot,D):
    idx = np.argsort(D)
    I = np.eye(len(D))
    return I[:,idx[:nroot]]

def update_R(r,e,D):
    for i in range(len(r)):
        r[i] *= 1/(e - D[i])
    return r

def orthogonalize_root(r,B):
    for i in range(B.shape[1]):
        b = B[:,i]/np.linalg.norm(B[:,i])
        r -= np.dot(b.T,r)*b
    return r

def davidson(A,D,nroot,nvec,maxit,tol,thresh_vec):

    vec_dim = A.shape[0]
    max_size = nroot*nvec

    B = diagonal_guess(nroot,D)
    curr_size = nroot

    for iter in range(maxit):

        B,_ = gramschmidt(B)

        sigma = np.dot(A,B)

        G = np.dot(B.T,sigma)
        
        # MUST USE np.linalg.eig NOT np.linalg.eigh
        eval_E, alpha = np.linalg.eig(G)

        idx = np.argsort(eval_E)
        eigval = eval_E[idx[:nroot]]
        alpha = alpha[:,idx[:nroot]]
        V = np.dot(B,alpha)

        add_B = np.zeros((vec_dim,nroot))
        resid_norm = np.zeros(nroot)
        ct_add = 0
        print('\nIter - {}    Subspace Dim = {}'.format(iter+1,curr_size))
        print('----------------------------------------------------')
        for j in range(nroot):
            r = np.dot(sigma,alpha[:,j]) - eigval[j]*V[:,j]
            resid_norm[j] = np.linalg.norm(r)
            q = update_R(r,eigval[j],D)
            q = q/np.linalg.norm(q)
            if ct_add > 0:
                q_orth = orthogonalize_root(q,np.concatenate((B,add_B[:,:ct_add]),axis=1))
            else:
                q_orth = orthogonalize_root(q,B)
            if np.linalg.norm(q_orth) > thresh_vec:
                add_B[:,ct_add] = q_orth/np.linalg.norm(q_orth)
                ct_add += 1

            print('Root - {}      e = {:0.8f}      |r| = {:0.8f}'.format(j+1,eigval[j],resid_norm[j]))

        if all(resid_norm <= tol):
            print('\nDavidson successfully converged!')
            break
        else:
            if curr_size >= max_size:
                print('Restarting and collapsing')
                B = np.dot(B,alpha)
                curr_size = nroot
            else:
                B = np.concatenate((B,np.asarray(add_B[:,:ct_add])),axis=1)
                curr_size = B.shape[1]

    return eigval, V
    
def main(args):
    
    n = args.size
    nvec = args.nvec
    tol = args.tolerance
    nroot = args.roots
    maxit = args.maxit
    thresh_vec = args.threshold

    sparsity = 10
    hermiticity = 0.5

    R = np.random.rand(n,n)
    A = np.diag(np.arange(1,n+1)) + sparsity*( hermiticity*R + (1.0-hermiticity)*R.T )
    D = np.diagonal(A)


    E_dav = np.zeros(nroot)
    V_dav = np.zeros((n,nroot))
    E_dav, V_dav = davidson(A,D,nroot,nvec,maxit,tol,thresh_vec)


    E_exact, V_exact = np.linalg.eig(A)
    idx = np.argsort(E_exact)
    E_exact = [E_exact[i] for i in idx]
    print('\n Root      Davidson         Exact')
    print('====================================')
    for i in range(nroot):
        print('  {}      {:0.8f}      {:0.8f}'.format(i+1,E_dav[i],E_exact[i]))


if __name__ == '__main__':

    parser = argparse.ArgumentParser()
    parser.add_argument('-v','--nvec',type=int,help='Max number of vectors per root',default=5)
    parser.add_argument('-t','--tolerance',type=float,help='Eigenvector residuum tolerance',default=1e-07)
    parser.add_argument('-n','--size',type=int,help='N x N size of test matrix',default=100)
    parser.add_argument('-r','--roots',type=int,help='Number of roots',default=5)
    parser.add_argument('-m','--maxit',type=int,help='Maximum number of iterations',default=30)
    parser.add_argument('-th','--threshold',type=float,help='Subspace addition threshold',default=1e-04)
    parser.add_argument('-s','--sparsity',type=float,help='Sparsity of random test matrix',default=2)
    parser.add_argument('-y','--hermitian',type=float,help='Degree of hermiticity of test matrix',default=0.5)
    args = parser.parse_args()
    main(args)

File: Davidson_matlab_test/davidson_solver/davidson.m
function [ R, eigval, resid_norm] = davidson(Ax, D, nroot, opts)

% Block-Davidson algorithm for diagonalizing large sparse (non-Hermitian)
% matrices

% References:
% K Hirao and H Nakatusji, J. Comp. Phys. 45, 246-254 (1982)
% R Parrish, E Hohenstein, and T J Martinez, J. Chem. Theory Comput, 12,
% 3003-3007 (2016)

% Input:
% Ax - matrix-vector product function for matrix A to be diagonalized
% D - diagonal of matrix A
% nroot - number of desired eigenpairs
% B0 - initial search space matrix, used in conjunction with
%      opts.init_guess = 'custom'
% opts - struct with fields
%        opts.maxit = maximum # of iterations (e.g. 100)
%        opts.tol = convergence tolerance for each root (e.g. 1e-5)
%        opts.max_nvec_per_root = maximum number of search vectors for each
%        root in subspace (e.g. 10)
%        opts.flag_verbose = print verbosity, 1 or 0

% Output
% V - right eigenvectors
% e - eigenvalues
% res - residuals for each root
% flag_conv - 1 if converged, 0 if not converged, 2 if stagnated


    nvec = opts.nvec;
    maxit = opts.maxit;
    tol = opts.tol;
    thresh_vec = opts.thresh_vec;

    fprintf('Beginning Davidson diagonlization algorithm...\n')

    
    tic
    
    
    %%%%%%% SOLVE RIGHT EIGENPROBLEM %%%%%%%%%
       

    max_size = nvec*nroot;
    vec_dim = length(D);
    
    e = zeros(nroot,1); 
    
    % diagonal guess for trial space
    [~,idx_diag] = sort(D,'ascend');
    Id = eye(vec_dim);
    B0 = Id(:,idx_diag(1:nroot));
    
    B = zeros(vec_dim,max_size);
    SIGMA = zeros(vec_dim,max_size);
    B(:,1:nroot) = B0;
        
    curr_size = nroot;
    for it = 1:maxit

        if it == 1
        	n_prev = 1;
        else
            n_prev = curr_size - ct_add + 1;
        end

        fprintf('\nIter-%d    Subspace size - %d\n',it,curr_size)
        fprintf('-------------------------------------------------------\n')
        
        B0 = mgson(B(:,1:curr_size));
        %B0 = gramschmidt(B(:,1:curr_size));
        B(:,1:curr_size) = B0;
        
        sigma = feval(Ax,B(:,n_prev:curr_size));
        SIGMA(:,n_prev:curr_size) = sigma;
        
        G = B(:,1:curr_size)'*SIGMA(:,1:curr_size);
        [alpha,GD] = eig(G);
        [e,idx] = sort(diag(GD),'ascend');

        % Right ritz vectors
        V = B(:,1:curr_size)*alpha;

        % sorting eigenpairs
        e = e(1:nroot);
        alpha = alpha(:,idx(1:nroot));  % expansion coefficients
        V = V(:,idx(1:nroot));

        % vectorized residual calculation
        RES = SIGMA(:,1:curr_size)*alpha - V*diag(e);
        resid_norm = sqrt(sum(RES.^2,1));
        idx_unc = find(resid_norm > tol);
        
        % printing 
        for j = 1:nroot
                fprintf('   Root - %d     e = %4.10f     |r| = %4.10f\n',j,real(e(j)),resid_norm(j));
        end

        add_B = zeros(vec_dim,nroot); ct_add = 0;
        for j = 1:length(idx_unc)
            
            J = idx_unc(j);

            RES(:,J) = update_R(RES(:,J),e(J),D);
            RES(:,J) = RES(:,J)/norm(RES(:,J));
            
            if ct_add > 0
                r_orth = orthogonalize_root(RES(:,J),[B(:,1:curr_size),add_B(:,1:ct_add)]);
            else
                r_orth = orthogonalize_root(RES(:,J),B(:,1:curr_size));
            end
            
            if norm(r_orth)/norm(RES(:,J)) > thresh_vec
                ct_add = ct_add + 1;
                add_B(:,ct_add) = r_orth/norm(r_orth);
            end

        end

        if all(resid_norm <= tol)
            R = V; eigval = e;
            fprintf('\nDavidson successfully converged in %d iterations (%4.2f seconds)\n',it, toc);
            return
        else
            if curr_size >= max_size
                fprintf('\nRestarting and collapsing...\n')
                B = B*alpha;
                curr_size = nroot;
            else
                B(:,curr_size+1:curr_size+ct_add) = add_B(:,1:ct_add);
                curr_size = curr_size + ct_add;
            end
        end


    end

    fprintf('\nDavidson failed to converge in %d iterations',it)
    R = V; eigval = e; 
        
end

